docs/2019-08-27-Unification-modulo-overloaded-constructors.md

```diff
@@ -0,0 +1,206 @@
+
+Problems
+--------
+
+Although overloading introduces a higher degree of modularization through
+some form of subtype-polymorphism, it puts a higher burden on the
+unification/matching procedure.
+
+1. Due to axioms like the above being generated, the overloaded constructors
+   are no longer proper constructors and therefore are not marked as such,
+   which hinders unification/matching, given that usually the lhs of a rule is
+   expected to be constructor-based. This is especially bad because it also
+   makes disproving of unification harder and thus generates more (im)possible
+   constrainted solutions.
+
+2. Unification of these constructs needs to be performed modulo overloading
+   axioms like the ones above.
+
+Proposed solution
+-----------------
+
+First observation is that although overloaded constructors are not proper
+constructors, they behave as such in most instances: they are injective, so
+no-confusion-same-constructor applies naturally, but also they are subject to
+most no-confusion-different-constructors rules.  Indeed, due to the
+__SingleOverloadPerSort__ constraint, the only confusion possible for an
+oveloaded operation is that with an injection operation as given by the
+overloading axioms.
+
+### Decision 1
+
+Hence, the proposed solution for handling problem (1) specified above is to
+detect overloaded constructors from the `overload` annotations associated to the
+overloading axioms and to and use that information in the unification/matching
+algorithm, handling overloaded constructors as constructor-like symbols.
+
+For the second problem, the issue comes from the fact that the unifying modulo
+the overloading axioms might require applying them either way (left-to-right or
+right-to-left).
+
+One advantage of left-to-right application of overloading axioms is that it
+helps computing the lowest sort for a term.
+
+### Example
+
+For example, consider lists of
+expressions overloaded by lists of values as in
+[SIMPLE](https://github.com/kframework/k/blob/master/k-distribution/tutorial/2_languages/1_simple/1_untyped/simple-untyped.k):
+
+```
+  syntax Exps ::= List{Exp,","}          [strict]  // automatically hybrid now
+  syntax Vals ::= List{Val,","}
+```
+
+Consider the definition of `Val`ues (including `lambda`), the syntax for
+function call, and the function call rule itself:
+
+```
+  syntax Exp ::= Exp "(" Exps ")"        [strict]
+
+  syntax Val ::= Int | Bool | String
+               | array(Int,Int)
+               | lambda(Ids,Stmt)
+  syntax Exp ::= Val
+  syntax Exps ::= Vals
+
+  rule <k> lambda(Xs,S)(Vs:Vals) ~> K => mkDecls(Xs,Vs) S return; </k>
+       <control>
+         <fstack> .List => ListItem((Env,K,C)) ...</fstack>
+         C
+       </control>
+       <env> Env => GEnv </env>
+       <genv> GEnv </genv>
+```
+
+The rule above requires that the second argument of the application expression
+(which should be a list of expressions, is actually a list of values).
```

Comment:
This parenthesis seems wrong. The open parenthesis should probably be a comma, the closing one should probably be removed.